File: Tests/AINotebookCoreTests/PlainTextExtractorTests.swift
import XCTest
@testable import AINotebookCore

final class PlainTextExtractorTests: XCTestCase {

    func testExtractsUtf8Plaintext() async throws {
        let url = try writeTempFile(name: "memo.txt", bytes: Data("Hello, world.".utf8))
        defer { try? FileManager.default.removeItem(at: url) }
        let extracted = try await PlainTextExtractor().extract(from: url, kind: .text)
        XCTAssertEqual(extracted.text, "Hello, world.")
        XCTAssertEqual(extracted.title, "memo")
    }

    func testStripsMarkdownLeadingHashes() async throws {
        let md = "# Title\n\nSome **bold** body."
        let url = try writeTempFile(name: "doc.md", bytes: Data(md.utf8))
        defer { try? FileManager.default.removeItem(at: url) }
        let extracted = try await PlainTextExtractor().extract(from: url, kind: .markdown)
        // Title is the first Markdown heading.
        XCTAssertEqual(extracted.title, "Title")
        // Markdown body retained (we do NOT lose content — we just expose
        // the raw text).
        XCTAssertTrue(extracted.text.contains("Some **bold** body."))
    }

    func testEmptyFileThrows() async throws {
        let url = try writeTempFile(name: "empty.txt", bytes: Data())
        defer { try? FileManager.default.removeItem(at: url) }
        do {
            _ = try await PlainTextExtractor().extract(from: url, kind: .text)
            XCTFail("expected throw")
        } catch ExtractorError.emptyContent {
            // ok
        } catch {
            XCTFail("wrong error: \(error)")
        }
    }

    private func writeTempFile(name: String, bytes: Data) throws -> URL {
        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("ai-notebook-tests-\(UUID().uuidString)")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let url = dir.appendingPathComponent(name)
        try bytes.write(to: url)
        return url
    }
}

File: windows/tests/AINotebook.Core.Tests/Extractors/PlainTextExtractorTests.cs
using System.Text;
using AINotebook.Core.Extractors;
using AINotebook.Core.Models;
using Xunit;

namespace AINotebook.Core.Tests.Extractors;

public class PlainTextExtractorTests : IDisposable
{
    private readonly string _dir;

    public PlainTextExtractorTests()
    {
        _dir = Path.Combine(Path.GetTempPath(), "ai-notebook-tests-" + Guid.NewGuid());
        Directory.CreateDirectory(_dir);
    }

    public void Dispose()
    {
        try { Directory.Delete(_dir, recursive: true); } catch { /* ignore */ }
    }

    private Uri WriteTempFile(string name, byte[] bytes)
    {
        var path = Path.Combine(_dir, name);
        File.WriteAllBytes(path, bytes);
        return new Uri(path);
    }

    [Fact]
    public async Task ExtractsUtf8Plaintext()
    {
        var url = WriteTempFile("memo.txt", Encoding.UTF8.GetBytes("Hello, world."));
        var extracted = await new PlainTextExtractor().ExtractAsync(url, SourceType.Text);
        Assert.Equal("Hello, world.", extracted.Text);
        Assert.Equal("memo", extracted.Title);
    }

    [Fact]
    public async Task StripsMarkdownLeadingHashes()
    {
        const string md = "# Title\n\nSome **bold** body.";
        var url = WriteTempFile("doc.md", Encoding.UTF8.GetBytes(md));
        var extracted = await new PlainTextExtractor().ExtractAsync(url, SourceType.Markdown);
        // Title is the first Markdown heading.
        Assert.Equal("Title", extracted.Title);
        // Markdown body retained (raw text exposed, markup not stripped).
        Assert.Contains("Some **bold** body.", extracted.Text);
    }

    [Fact]
    public async Task EmptyFileThrows()
    {
        var url = WriteTempFile("empty.txt", Array.Empty<byte>());
        await Assert.ThrowsAsync<ExtractorException.EmptyContent>(
            () => new PlainTextExtractor().ExtractAsync(url, SourceType.Text));
    }
}
